### PR TITLE
feat: Add leader receiver creator receiver

### PR DESCRIPTION
**Pull request by rakesh-garimella**
**Description**

Changes proposed in this pull request:

- Add singleton receiver creator receiver which wraps a sub-receiver. 
- Add unit tests to provide test coverage
-  Added Readme
- Read leader election config from the configmap

**Related issue(s)**

https://github.com/kyma-project/opentelemetry-collector-components/issues/7

### CODE REVIEW

File: receiver/leaderreceivercreator/config.go [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+func newLeaderElectionConfig(lecConfig leaderElectionConfig, cfg map[string]any) (leaderElectionConfig, error) {
+	if leaseName, ok := cfg["lease_name"].(string); ok {
+		lecConfig.leaseName = leaseName
+	}
+	if leaseNamespace, ok := cfg["lease_namespace"].(string); ok {
+		lecConfig.leaseNamespace = leaseNamespace
+	}
+
+	if leaseDuration, ok := cfg["lease_duration"].(string); ok {
+		fmt.Printf("leaseDuration: %v\n", leaseDuration)
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply receiver/leaderreceivercreator/config.go]: removed

File: receiver/singletonreceivercreator/README.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        protocols:
+          grpc:
+            endpoint: ""
+```
+The configuration consists of two parts:
+1. The leader election configuration.
+2. The receiver configuration.
+
+### Leader Election Configuration
+| configuration   | description                                                                |
```
Comment (chrkl) [inline receiver/singletonreceivercreator/README.md:41]: Maybe also describe the defaults

File: receiver/singletonreceivercreator/leader_elector.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+	"context"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+)
+
+const (
+	inClusterNamespacePath = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
```
Comment (chrkl) [inline receiver/singletonreceivercreator/leader_elector.go:15]: Where is this used?

File: receiver/singletonreceivercreator/receiver.go [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+// Start leader receiver creator.
+func (c *singletonReceiverCreator) Start(_ context.Context, host component.Host) error {
+	c.host = host
+	// Create a new context as specified in the interface documentation
+	ctx := context.Background()
+	ctx, c.cancel = context.WithCancel(ctx)
+
+	c.params.TelemetrySettings.Logger.Debug("Starting leader election receiver...")
```
Comment (chrkl) [inline receiver/singletonreceivercreator/receiver.go:48]: Which log level do other receivers for such messages? Ensure to be consistent. Use consistent naming for the receiver.

File: receiver/singletonreceivercreator/receiver_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/component/componenttest"
+	"go.opentelemetry.io/collector/receiver/receivertest"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/fake"
+)
+
+func TestMockReceiverCreator(t *testing.T) {
```
Comment (chrkl) [inline receiver/singletonreceivercreator/receiver_test.go:20]: Adapt name

File: receiver/singletonreceivercreator/factory.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
+// Copyright The OpenTelemetry Authors
```
Comment (chrkl) [inline receiver/singletonreceivercreator/factory.go:1]: You use these headers inconsistently. If you add any copyright claims, make them proper. If any of the files have been copied from other Otel collector repos, make that clear and add the source.

File: receiver/singletonreceivercreator/deploy/kube/collectors-no-leader.yaml [hunk 7]
```diff
@@ -0,0 +1,69 @@
+---
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply receiver/singletonreceivercreator/deploy/kube/collectors-no-leader.yaml:1]: I removed this file as it makes no sense to have deployment without leader election

File: receiver/singletonreceivercreator/deploy/kube/rbac.yaml [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+  - create
+  - update
+  - patch
+  - delete
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: cluster-monitor
+rules:
```
[the remark this answers is not in the document]
Comment (rakesh-garimella) [reply receiver/singletonreceivercreator/deploy/kube/rbac.yaml:25]: The cluster role is required for k8s cluster reciever: check here: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/k8sclusterreceiver/testdata/e2e/collector/clusterrole.yaml

**Review by chrkl — DISMISSED** [review]

**Review by chrkl — DISMISSED** [review]

File: receiver/singletonreceivercreator/README.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
+# Singleton Receiver Creator
+
+| Status      |                            |
+|-------------|----------------------------|
+| stability   | alpha: metrics             |
+| Code Owners | kyma-project/observability |
+
+Singleton Receiver Creator is a OTel Collector receiver that instantiates another receiver based on the leader election status. It is useful when one wants to have a single instance of a receiver running in a cluster. The receiver which gets the lease is created and executed.
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:8]:
```suggestion
Singleton Receiver Creator is an OTel Collector receiver that instantiates another receiver based on the leader election status. It is useful when you want to have a single instance of a receiver running in a cluster. The receiver that gets the lease is created and executed.
```

File: receiver/singletonreceivercreator/README.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+The problem this implementation solves has been mentioned in following [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32994). In future this implementation can be used as solution to the aforementioned issue.
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:9]:
```suggestion
This implementation solves the problem mentioned in the following [issue](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/32994). In the future, this implementation can be used as a solution to the aforementioned issue.
```

File: receiver/singletonreceivercreator/README.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+
+## How it works
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:11]:
```suggestion
## How It Works
```

File: receiver/singletonreceivercreator/README.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+It utilizes leader election to determine which instance should be the leader. The instance which wins the lease, becomes the leader and starts the underlying sub-receiver. When the leader loses the lease, it will stop the receiver and wait until it acquires the lease again.
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:13]:
```suggestion
It utilizes leader election to determine which instance should be the leader. The instance that wins the lease becomes the leader and starts the underlying sub-receiver. When the leader loses the lease, it stops the receiver and waits until it acquires the lease again.
```

File: receiver/singletonreceivercreator/README.md [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+| configuration   | description                                                                | default value      |
+|-----------------|----------------------------------------------------------------------------|--------------------|
+| lease_name      | The name of the lease object.                                              | singleton-receiver |
+| lease_namespace | The namespace of the lease object.                                         | default            |
+| lease_duration  | The duration of the lease.                                                 | 15s                |
+| renew_deadline  | The deadline for renewing the lease. It should be less than lease duration | 10s                |
+| retry_period    | The period for retrying the leader election.                               | 2s                 |
+
+### Receiver Configuration
+The `name` field specifies the name of the receiver that needs to created when the instance becomes the leader, followed by the configuration of the receiver that needs to be created.
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:49]:
```suggestion
The **name** field specifies the name of the receiver that needs to be created when the instance becomes the leader, followed by the configuration of the receiver that needs to be created.
```

File: receiver/singletonreceivercreator/README.md [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+
+
+### Multiple receiver with Singleton Receiver Creator
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:52]:
```suggestion
### Multiple Receivers with Singleton Receiver Creator
```

File: receiver/singletonreceivercreator/README.md [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+For leveraging singleton receiver creator with multiple sub receiver, one needs to have to repeat the leader election configuration for each receiver. This is a conscious design decision, where each receiver would have to create and acquire its own lease resource. The main motivation for the decision is that we can divide the load of the pod where the receiver is running as running multiple receivers on same pod would increase the load on the pod.
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:53]:
```suggestion
To leverage Singleton Receiver Creator with multiple sub-receivers, you must repeat the leader election configuration for each receiver. This is a conscious design decision where each receiver must create and acquire its own lease resource. The main motivation for the decision is that we can divide the load of the Pod where the receiver is running, as running multiple receivers on the same Pod would increase the load on the Pod.
```

File: receiver/singletonreceivercreator/README.md [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+
+
+## How to test
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:56]:
```suggestion
## How to Test
```

File: receiver/singletonreceivercreator/README.md [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+2. The receiver configuration.
+
+### Leader Election Configuration
+| configuration   | description                                                                | default value      |
+|-----------------|----------------------------------------------------------------------------|--------------------|
+| lease_name      | The name of the lease object.                                              | singleton-receiver |
+| lease_namespace | The namespace of the lease object.                                         | default            |
+| lease_duration  | The duration of the lease.                                                 | 15s                |
+| renew_deadline  | The deadline for renewing the lease. It should be less than lease duration | 10s                |
+| retry_period    | The period for retrying the leader election.                               | 2s                 |
```
Comment (IwonaLanger) [inline receiver/singletonreceivercreator/README.md:46]:
```suggestion
|-----------------|----------------------------------------------------------------------------|--------------------|
| **lease_name**      | The name of the lease object.                                              | singleton-receiver |
| **lease_namespace** | The namespace of the lease object.                                         | default            |
| **lease_duration**  | The duration of the lease.                                                 | 15s                |
| **renew_deadline**  | The deadline for renewing the lease. It should be less than the lease duration. | 10s                |
| **retry_period**    | The period for retrying the leader election.                               | 2s                 |
```

**Review by IwonaLanger — DISMISSED** [review]

**Review by IwonaLanger — APPROVED** [review]

**Pull request merged**